_pages/algorithm-development/advanced-algorithm-development/dealing-with-bottlenecks.md

```diff
@@ -0,0 +1,356 @@
+import numpy as np
+
+client = Algorithmia.client()
+graph, label_index = load_model(client)
+
+# API calls will begin at the apply() method, with the request body passed as 'input'
+# For more details, see algorithmia.com/developers/algorithm-development/languages
+    
+def get_image(url):
+    """Uses the Smart Image Downloader algorithm to format and download images from the web or other places."""
+    input = {'image': str(url)}
+    output_url = client.algo("util/SmartImageDownloader/0.2.x").pipe(input).result['savePath'][0]
+    temp_file = client.file(output_url).getFile().name
+    os.rename(temp_file, temp_file + '.' + output_url.split('.')[-1])
+    return temp_file + '.' + output_url.split('.')[-1]
+
+
+def do_work(image):
+    """Does some computer vision work and needs a numpy array to function"""
+    image_data = tf.gfile.FastGFile(image, 'rb').read()
+    with tf.Session(graph=graph) as sess:
+        softmax_tensor = sess.graph.get_tensor_by_name('softmax:0')
+        predictions = sess.run(softmax_tensor,
+                               {'DecodeJpeg/contents:0': image_data})
+        predictions = np.squeeze(predictions)
+        tags = []
+        top_k = predictions.argsort()[-5:][::-1]
+
+        for node_id in top_k:
+            human_string = id_to_string(label_index, node_id)
+            score = predictions[node_id]
+            result = {}
+            result['class'] = human_string
+            result['confidence'] = score.item()
+            tags.append(result)
+        results = {}
+        results['tags'] = tags
+        return results
+
+def apply(input):
+    image_data = get_image(input)
+    results = do_work(image_data)
+    return results
+
+```
+
+This is works perfectly for situations where you just want to get image support to an algorithm that needs to do serial processing and then do some monolithic processing using a gpu or other resources to get a final result, one image at a time.
+However, if you're dealing with a production system, you're most likely going to be using more than 1 downstream algorithm; and you'll also want to improve performance as much as possible, especially in batch. In that case, you're in luck! This tutorial will go over the different mechanisms you can use both inside of your algorithm, and outside to get the results you're looking for.
+
+# Bandwidth Bottlenecks
+
+## Async and Futures
+Your algorithm above works great, but now you need to improve performance and enable batch processing. 
+Your algorithm executes code very quickly (Which is great!) but the http reqeuests to actually download images or some other web resource take a variable amount of time, and are eating up a huge chunk of the total compute time.
+Well we're in luck; having multiple threads open trying to access http/https resources don't consume precious cpu clock cycles due to the [DMA](https://en.wikipedia.org/wiki/Direct_memory_access) controllers built into all modern CPUs.
```

Comment:
no more semi-colons :-) those are for people who can't commit to either a period or a comma lol

Reply:
LOL I'm gonna disagree on this one... author is adopting a one-sentence-per-line motif and semicolon is a prefect way to do it. Myself, I'd use a colon tho :-p

Reply:
I'm a big semi-colon fan as it emposes a different kind of pause to the reader, it's more of a suggested pause; rather than a hard one.